docs:#600two stage bootstrap (#700)
redpanda-data/documentation: docs/deployment/production-deployment.mdx

:::

## Configure and start seed nodes

Configure and start up the seed servers with the [`rpk config bootstrap`](../../reference/rpk/rpk-redpanda/rpk-redpanda-config-bootstrap) command: 

```bash
sudo rpk config bootstrap --self <private ip> --ips <seed nodes ips> \ 
sudo systemctl start redpanda-tuner redpanda
```

> - The `--self` flag tells the node which interface address to bind to. Usually you want that to be its private IP.

Reviewer: ```suggestion
- The `--self` flag tells the node which interface address to bind to. Usually this is its private IP.
```

- The `--ips` flag must be identical on all nodes when deploying a cluster for the first time.

> A Redpanda cluster starts with the seed nodes that were supplied to the `--ips` flag, instantiating a controller Raft group with the seed nodes. The cluster grows as nodes start up and is available after all seed servers are up. 

Reviewer: ```suggestion
A Redpanda cluster starts with the seed nodes that are specified in the `--ips` flag, instantiating a controller Raft group with the seed nodes. The cluster grows as nodes start up and is available after all seed servers are up. 
```

:::info
- Redpanda strongly recommends at least three seed nodes when forming a cluster. A larger number of seed nodes increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data.
- It's important to have one or more seed nodes in each fault domain (such as rack or cloud AZ). A higher number provides a stronger guarantee that clusters don’t fracture unintentionally.

> - It's possible to change the seed nodes for a short period of time after a cluster has been created. For example, you may want to designate one more node as a seed node to increase availability. To do that without cluster downtime, add the new node to [`seed_servers`](../../reference/node-properties/) and restart Redpanda to apply the change on a node-by-node basis.

Reviewer: ```suggestion
- It's possible to change the seed nodes for a short period of time after a cluster has been created. For example, you may want to designate one more nodes as a seed node to increase availability. To do this without cluster downtime, add the new node to [`seed_servers`](../../reference/node-properties/) and restart Redpanda to apply the change on a node-by-node basis.
```

Author: fixed as: "... designate one additional node as a seed node to increase availability. To do this ..."

- The `--ips` flag must be identical on all nodes when deploying a cluster for the first time.

:::info
- Redpanda strongly recommends at least three seed nodes when forming a cluster. A larger number of seed nodes increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data.
- It's important to have one or more seed nodes in each fault domain (such as rack or cloud AZ). A higher number provides a stronger guarantee that clusters don’t fracture unintentionally.
- It's possible to change the seed nodes for a short period of time after a cluster has been created. For example, you may want to designate one more node as a seed node to increase availability. To do that without cluster downtime, add the new node to [`seed_servers`](../../reference/node-properties/) and restart Redpanda to apply the change on a node-by-node basis.
:::

The cluster is available after all seed servers are up. 

If clients will connect from a different subnet, see [Configuring Listeners](/docs/cluster-administration/listener-configuration).

## Configure and start other nodes

> Let every other node know where to reach seed nodes:

Author: fixed

```bash
sudo rpk config bootstrap --self <private ip> --ips <seed nodes ips> \ 
sudo systemctl start redpanda-tuner redpanda
```

> A Redpanda cluster starts with the seed nodes that were supplied to the `--ips` flag, instantiating a controller Raft group with the seed nodes. The cluster grows as nodes start up. 

Author: Thanks for the close review @andrwng! See: https://deploy-preview-700--redpanda-documentation.netlify.app/docs/deployment/production-deployment/#configure-and-start-seed-nodes

:::

## Configure and start seed nodes

Configure and start up the seed servers with the [`rpk config bootstrap`](../../reference/rpk/rpk-redpanda/rpk-redpanda-config-bootstrap) command: 

```bash

> sudo rpk config bootstrap --self --ips <hostnames of seed nodes> \

Author: fixed

sudo systemctl start redpanda-tuner redpanda
```

The `--self` flag tells the node which interface address to bind to. Usually you want that to be its private IP.

:::info
When deploying a cluster for the first time, the `--ips` flag must be identical on all nodes.
:::

If clients will connect from a different subnet, see [Configuring Listeners](/docs/cluster-administration/listener-configuration).

## Configure and start other nodes

> Let every other node know where to reach the root node:

Author: fixed

```bash
sudo rpk config bootstrap \
--self <private ip>                        \

> --ips <root node ip> &&                    \

Author: fixed

sudo systemctl start redpanda-tuner redpanda
```

> Redpanda starts with three configurable nodes, instantiating a controller Raft group with the seed nodes. The cluster grows as nodes start up.

Author: fixed

:::info
- Redpanda strongly recommends at least three seed nodes when forming a cluster. A larger number of seed nodes increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data.
- It's important to have one of more seed nodes in each fault domain (such as rack or cloud AZ). A higher number provides a stronger guarantee that clusters don’t fracture unintentionally.
:::

> Redpanda automatically generates unique node IDs for each new node. You can choose to assign node IDs in configuration files, but as a part of bootstrapping a cluster, Redpanda automatically creates node IDs and enforces that they have unique node IDs. 

Author: Rewrote as: 
"A Redpanda cluster starts with the seed nodes that were supplied to the --ips flag, instantiating a controller Raft group with the seed nodes. The cluster grows as nodes start up. As a part of bootstrapping a cluster, Redpanda automatically generates unique values for each new node."

:::note
Previous versions of Redpanda required making sure that each node had a unique `node_id` set in its configuration file and that exactly one node had an empty `seed_servers` set in its configuration file. A single (root) node instantiated the cluster, and other nodes joined it. For information about doing this today, see [Override default to start seed nodes](../../deployment/custom-deployment/#override-default-to-start-seed-nodes).

> :::

Author: removed!

:::

> ## Configure and start seed nodes

Reviewer 2: Can we add a warning or note mentioning that when deploying a cluster for the first time, it is critical for the `--ips` flag to be identical on all nodes?

Author: added

> Configure and start the root node with the `rpk config bootstrap` command. The `--self` option tells the node which interface address to bind to. Usually you want that to be its private IP.

Author: Thank you! Fixed.

```bash

> sudo rpk config bootstrap --self <ip> && \

Reviewer 2: We need to include something like the following, here and below (instead of "root node ip").

```
--ips <hostnames of seed nodes>
```

Author: @andrwng is this correct:
sudo rpk config bootstrap --self --ips <hostnames of seed nodes> \

Author: now says: sudo rpk config bootstrap --self <ip> --ips <seed nodes ips>

:::

#### Optional: Manually instantiate cluster

Older versions of Redpanda required manual instantiation. For backward compatibility, Redpanda still support this. It requires making sure that each node has a unique `node_id` set in its configuration file and exactly one node has an empty `seed_servers` set in its configuration file. A single (root) node instantiates the cluster, and other nodes join it.  
## Configure and start seed nodes

First configure and start the root node. For the root node, choose 0 as its ID. The `--self` option tells the node which interface address to bind to. Usually you want that to be its private IP.
Configure and start the root node with the `rpk config bootstrap` command. The `--self` option tells the node which interface address to bind to. Usually you want that to be its private IP.

```bash
sudo rpk config bootstrap --id 0 --self <ip> && \
sudo rpk config bootstrap --self <ip> && \
sudo systemctl start redpanda-tuner redpanda
```

If clients will connect from a different subnet, see [Configuring Listeners](../../cluster-administration/listener-configuration).

> ### Configure and start other nodes

Reviewer 2: Why'd we get rid of the "start other nodes" line? I thought it was a good bisection of the steps:
- start the seeds
- start the other nodes

Author: @andrwng I put this back in, but I'm confused by this. The example seems the same for both start the seeds and start the other nodes. Will you please review the draft again?

Reviewer 3: I've no problems with starting seeds and joiners simultaneously. However the docs may benefit from describing how to configure them separately, e.g. focusing on selection of available set of seeds in the seeds-related section, and discuss that joiners may have different `seed_servers` if required in the joiners-related section.

Reviewer 2: >discuss that joiners may have different seed_servers if required in the joiners-related section

Is that something we want to advertise? I worry having different configurations will be confusing.

Reviewer 3: That may be important to be able to add/decommission seed nodes without the entire cluster downtime. I think if it is not mentioned, people may think from the docs that the only way to change the seed nodes set is through cluster downtime, which is not good. So maybe we can add a paraprah like this:

---

Although it is important for cluster health to keep the list of `seed_servers` identical in every node in the cluster, there is no requirement for that once a cluster has been created. It is ok to have the `seed_servers` lists to be different for a short period of time.

One example where that may be needed is in case you want to designate one more node as a seed node to increase seeds availability. Do do that without cluster downtime, you can add the new item to `seed_servers` and restart Redpanda to apply the change on node-by-node basis.

Author: @dlex @andrwng I added a bullet to the list for this, but still have 2 separate sections (with start other nodes). Is this OK: https://deploy-preview-700--redpanda-documentation.netlify.app/docs/deployment/production-deployment/#configure-and-start-seed-nodes

### Seed servers list

The `seed_servers` node configuration property controls how Redpanda finds its peers when initially forming a cluster. Redpanda starts nodes with a non-empty `seed_servers` list, and it sends requests to the nodes in that list to join the cluster.

> All nodes can have identical `seed_servers` in their configuration files. 

Author: @andrwng: Please let me know what you think about this updated draft: https://deploy-preview-700--redpanda-documentation.netlify.app/docs/deployment/production-deployment/#configure-and-start-other-nodes

:::note Notes

> - The minimum number of seed servers to form a cluster is three, but you can adjust this to increase the robustness of consensus and minimize any chance that new clusters get spuriously formed after nodes are lost or restarted without any data. 

Reviewer 2: This isn't quite correct. How about:

"It is strongly recommended that there be at least three seed servers when forming a cluster. A larger number of seed servers increases the robustness of consensus and minimizes any chance that new clusters get spuriously formed after nodes are lost or restarted without any data."

Author: Thank you. Incorporated this.

:::

### Instantiate a cluster

The `minimum_initial_cluster_size` bootstrap configuration property is set to `1` by default. For production deployments, set this to `3`. This enables a simpler configuration of node IDs and seed servers.

#### Set cluster size to three

With `minimum_initial_cluster_size=3`, clusters start with three configurable nodes, instantiating a controller Raft group with the seed servers list. The cluster grows as nodes start up. 

> Redpanda automatically generates node IDs, guaranteeing that new nodes get unique node IDs. Nodes assign themselves a UUID (persisted in their kv-store), and register themselves with the controller leader, who assigns a uniue node ID. This means that you don't need to include node IDs in configuration files or worry about policies on `node_id` re-use. Redpanda initially starts up without an ID, and a node gets an ID only after it joins the cluster. All nodes can have identical `seed_servers` in their configuration files. 

Reviewer 4: ```suggestion
Redpanda automatically generates unique node IDs for each new node. Nodes assign themselves a UUID (persisted in their kv-store), and register themselves with the controller leader, who assigns a unique node ID. This means that you don't need to include node IDs in configuration files or worry about policies on `node_id` re-use. Redpanda initially starts up without an ID, and a node gets an ID only after it joins the cluster. All nodes can have identical `seed_servers` in their configuration files. 
```

Author: thanks @AnneLeeperRP! This is still an early WIP, but this helps!
